pkg/controller/azdrivernode.go

```diff
@@ -88,25 +91,40 @@ func (r *ReconcileAzDriverNode) Reconcile(ctx context.Context, request reconcile
 
 // run an update on existing azdrivernode objects to store them under new version if necessary
 func (r *ReconcileAzDriverNode) Recover(ctx context.Context) error {
+	var errs []error
 	var err error
 	ctx, w := workflow.New(ctx)
 	defer func() { w.Finish(err) }()
 
 	var nodes *azdiskv1beta2.AzDriverNodeList
 	if nodes, err = r.azClient.DiskV1beta2().AzDriverNodes(r.objectNamespace).List(ctx, metav1.ListOptions{}); err != nil {
-		if errors.IsNotFound(err) {
-			return nil
-		}
 		return err
 	}
 
 	for _, node := range nodes.Items {
-		updated := node.DeepCopy()
-		updated.Annotations = azureutils.AddToMap(updated.Annotations, consts.RecoverAnnotation, "azDriverNode")
-		if _, err = r.azClient.DiskV1beta2().AzDriverNodes(r.objectNamespace).Update(ctx, updated, metav1.UpdateOptions{}); err != nil {
-			return err
+		// if the corresponding node has been deleted, delete the azdriverode object
+		_, err := r.kubeClient.CoreV1().Nodes().Get(ctx, node.Spec.NodeName, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			azNode := r.azClient.DiskV1beta2().AzDriverNodes(r.objectNamespace)
+			if err = azNode.Delete(ctx, node.Name, metav1.DeleteOptions{}); err != nil {
+				errs = append(errs, err)
+			}
+		} else {
+			updateFunc := func(obj client.Object) error {
+				azNode := obj.(*azdiskv1beta2.AzDriverNode)
+				azNode.Annotations = azureutils.AddToMap(azNode.Annotations, consts.RecoverAnnotation, "azDriverNode")
+				return nil
+			}
+			if _, err = azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, &node, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRI); err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			r.addNodeToAvailableAttachmentsMap(ctx, node.Name, node.GetLabels())
 		}
-		r.addNodeToAvailableAttachmentsMap(ctx, node.Name, node.GetLabels())
+	}
+	if len(errs) > 0 {
+		err = status.Errorf(codes.Internal, "azDriverNode recovery failed with: %+v", errs)
```

Comment:
Rather than aggregating all of the errors in one error return, we should log the individual node errors as they are encountered. If there are any failures, it's fine to return a generic error from this method. I am concerned that if there are 100s to 1000s of nodes, we may end up with one really large error log which is difficult to use for debugging.

Reply:
Good idea!